SSP Cache for Detail panel's Connections

bindeali opened the pull request:
## Info:
* Can be merged and closed right away.

## Changes:
* Implemented the SSP Cache for the Connections card in the Detail Panel.
* Fixed an issue where the Expand Connections button (➕) was not saving the results.
* Fixed an issue where selecting both workspace terms and cache search terms selected only those of either group, not both.
* Fixed the Cache button label ("Terms from other languages" -> "Terms outside the workspace").

## Deploy previews: 

* [Veřejné rejstříky](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-224258741)
* [eSbírka](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-1326236660)
* [Datový model OFN číselníků a statistických klasifikací](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-2083940323)
* [Karel test](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-415314821)
* [AB test](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance2011031447)
* [stenchlak-test](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-1654748901)
* [Turisticky cil](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance1044794056)
* [čas](https://deploy-preview-275--ontographer.netlify.app/?workspace=https://slovn%C3%ADk.gov.cz/datov%C3%BD/pracovn%C3%AD-prostor/pojem/metadatov%C3%BD-kontext/instance-202167279)

karelklima commented on src/function/FunctionLink.ts:
```diff
@@ -298,11 +306,179 @@ export function addLinkTools(
+    target: string;
+    labels: { [key: string]: string };
+  }[] = [];
+  const connections: {
```
It would be definitely better to specify complex types like this one outside of a function so that the types may be reused... and also it makes the function quite hard to read with so much going on.

karelklima commented on src/function/FunctionLink.ts:
```diff
@@ -298,11 +306,179 @@ export function addLinkTools(
+  }[] = [];
+
+  if (AppSettings.representation === Representation.FULL) {
+    await fetchRelationships(AppSettings.contextEndpoint, [iri], restrictions);
```
The "FULL" and "COMPACT" cases should be IMHO separated to standalone functions.

karelklima replied:
Also do I understand correclty that the `restrictions` array is passed to the `fetchRelationships` method that populates it? It would be much better if the function returned the respective array. In general modifying function arguments is something that should never be done, unless there is a good enough reason for that. It also helps readability.

karelklima commented on src/function/FunctionLink.ts:
```diff
@@ -298,11 +306,179 @@ export function addLinkTools(
+    await fetchRelationships(AppSettings.contextEndpoint, [iri], restrictions);
+    await fetchReadOnlyTerms(
+      AppSettings.contextEndpoint,
+      Object.keys(restrictions)
```
This array should be probably specified upfront and assigned to a variable before the await call, for the benefit of readability.

karelklima commented on src/function/FunctionLink.ts:
```diff
@@ -298,11 +306,179 @@ export function addLinkTools(
+    );
+    for (const restriction of WorkspaceTerms[iri].restrictions) {
+      if (restriction.target in terms) {
+        const newConnection = {
```
The newConnection construct should be extracted to a standalone function preferrably - there are multiple calls like this one in this function and all of them are almost the same.

karelklima commented on src/function/FunctionLink.ts:
```diff
@@ -298,11 +306,179 @@ export function addLinkTools(
+      relationships.map((rel) => rel.target).concat(subClasses),
+      terms
+    );
+    for (const relationship of relationships.filter(
```
This is very hard to read. The filtered relationships should be extracted to a variable before the for loop code. Similar pattern occurs on many places in this function.

karelklima commented on src/panels/detail/components/connections/ConnectionCache.tsx:
```diff
@@ -0,0 +1,133 @@
+import Connection from "./Connection";
+
+interface Props {
+  connection: {
```
This type is just copied from another file... should be definitely extracted to a separate external type so that it can be reused.

karelklima commented on src/panels/detail/components/connections/ConnectionCache.tsx:
```diff
@@ -0,0 +1,133 @@
+interface State {}
+
+export default class ConnectionCache extends React.Component<Props, State> {
+  checkVocabulary(vocabulary: string) {
```
What is this supposed to do? It is weird that a function named "check..." returns a non-boolean value.

karelklima commented on src/panels/detail/components/connections/ConnectionCache.tsx:
```diff
@@ -0,0 +1,133 @@
+          <Popover id={"termDetailPopover"}>
+            <Popover.Title as="h3">
+              {getLabelOrBlank(
+                this.props.connection.target.labels,
```
`const { connection, projectLanguage } = props;` at the beginning of the  render method would make the code much more readable.

karelklima commented on src/panels/detail/components/connections/ConnectionCache.tsx:
```diff
@@ -0,0 +1,133 @@
+        }
+      >
+        <Connection
+          onDragStart={(event: DragEvent) => {
```
This complex function should definitely be separated.

karelklima commented on src/panels/detail/components/connections/ConnectionList.tsx:
```diff
@@ -11,26 +11,44 @@ import {
   shownConnections: string[];
   showFilter: boolean;
+  showLucene: boolean;
+  shownLucene: {
```
Is this the same type as connection?

karelklima commented on src/queries/get/CacheQueries.ts:
```diff
@@ -67,40 +71,97 @@ export async function fetchVocabularies(
+  endpoint: string,
+  term: string
+): Promise<string[]> {
+  const query = [
```
Why didn't you use the SPARQL builder here?

karelklima commented on src/queries/get/CacheQueries.ts:
```diff
@@ -331,22 +396,85 @@ export async function fetchReadOnlyTerms(
               setSchemeColors(AppSettings.viewColorPool);
             }
-          }
+          } else if (sendTo) Object.assign(sendTo, result);
```
What does this line really do? It is not clear whether the result is returned at all.

karelklima reviewed (commented):
In general it's OK, but there are some issues. The most important ones:

1) Function are huge, should be broken down to smaller chunks.
2) There is a LOT of redundant code. More abstraction should be used.
3) Complex types should be defined outside of functions so that they can be reused.
4) Too complex code (e.g. function calls with callbacks) inside for loops or if conditions - in cases like that it's better to separate the code to a standalone variable.

bindeali commented:
Refactors done in e0bff03:
* The main loading function (the one that you levied criticism on the most) is split into multiple functions, such that reuse is maximized.
* Introduced a new Type for the function's connections.
* Fixed a typo in the English translation.
* Refactored the fetching functions such that they no longer have the responsibility of processing the results (i.e. if a new term comes from an as of yet unrecognized vocabulary, get the vocabulary and create a new "folder" with a new color - procedures such as these have been refactored into a new function). The "sendTo" procedures have therefore been eliminated.
